File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/query/QueriedFeature.java
package org.maplibre.nativeffi.query;

import java.util.Objects;
import java.util.Optional;
import org.maplibre.nativeffi.geo.Feature;
import org.maplibre.nativeffi.json.JsonValue;

/** One feature copied from a rendered or source feature query result. */
public record QueriedFeature(
    Feature feature,
    Optional<String> sourceId,
    Optional<String> sourceLayerId,
    Optional<JsonValue> state) {
  public QueriedFeature {
    Objects.requireNonNull(feature, "feature");
    sourceId = sourceId == null ? Optional.empty() : sourceId;
    sourceLayerId = sourceLayerId == null ? Optional.empty() : sourceLayerId;
    state = state == null ? Optional.empty() : state;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/query/QueriedFeature.kt
package org.maplibre.nativeffi.query

import org.maplibre.nativeffi.geo.Feature
import org.maplibre.nativeffi.json.JsonValue

/** One feature copied from a rendered or source feature query result. */
public data class QueriedFeature(
  public val feature: Feature,
  public val sourceId: String? = null,
  public val sourceLayerId: String? = null,
  public val state: JsonValue? = null,
)
